test/libsolutil/JSON.cpp

```diff
@@ -94,15 +146,42 @@ BOOST_AUTO_TEST_CASE(parse_json_strict)
 	BOOST_CHECK(json["3"]["3.1"] == "3.1");
 	BOOST_CHECK(json["3"]["3.2"] == 5);
 
-	// root element can only be object or array
+	// According to ECMA-404 object, array, number, string, true, false, null are allowed
+	// ... but JSONCPP disallows value types
 	BOOST_CHECK(jsonParseStrict("[]", json, &errors));
+	BOOST_CHECK(json.isArray());
 	BOOST_CHECK(jsonParseStrict("{}", json, &errors));
+	BOOST_CHECK(json.isObject());
 	BOOST_CHECK(!jsonParseStrict("1", json, &errors));
+	// BOOST_CHECK(json.isNumeric());
 	BOOST_CHECK(!jsonParseStrict("\"hello\"", json, &errors));
-
-	// non-UTF-8 escapes allowed??
+	// BOOST_CHECK(json.isString());
+	BOOST_CHECK(!jsonParseStrict("true", json, &errors));
+	// BOOST_CHECK(json.isBool());
+	BOOST_CHECK(!jsonParseStrict("null", json, &errors));
+	// BOOST_CHECK(json.isNull());
+
+	// Single quotes are also disallowed by ECMA-404
+	BOOST_CHECK(!jsonParseStrict("'hello'", json, &errors));
+	// BOOST_CHECK(json.isString());
+
+	// Only string keys in objects are allowed in ECMA-404
+	BOOST_CHECK(!jsonParseStrict("{ 42: \"hello\" }", json, &errors));
+
+	// According to ECMA-404 hex escape sequences are not allowed, only unicode (\uNNNN) and
+	// a few control characters (\b, \f, \n, \r, \t)
+	//
+	// More lenient parsers allow hex escapes as long as they translate to a valid UTF-8 encoding.
+	//
+	// ... but JSONCPP allows any hex escapes
 	BOOST_CHECK(jsonParseStrict("[ \"\x80\xec\x80\" ]", json, &errors));
+	BOOST_CHECK(json.isArray());
 	BOOST_CHECK(json[0] == "\x80\xec\x80");
+
+	// This would be valid more lenient parsers.
+	BOOST_CHECK(jsonParseStrict("[ \"\xF0\x9F\x98\x8A\" ]", json, &errors));
```

Comment:
I'd add comments showing what these escapes are meant to decode into:
```suggestion
	BOOST_CHECK(jsonParseStrict("[ \"\x80\xec\x80\" ]", json, &errors)); // [ [ "ì" ] ]
	BOOST_CHECK(json.isArray());
	BOOST_CHECK(json[0] == "\x80\xec\x80");

	// This would be valid more lenient parsers.
	BOOST_CHECK(jsonParseStrict("[ \"\xF0\x9F\x98\x8A\" ]", json, &errors)); // [ [ "ð" ] ]
```

Reply:
What is `"ð"` ?

The unicode is for the smiley emoji :)

Reply:
Oh :) That's what I got when I printed the string in Python. But now I realized that I should have used a bytestring and decoded it - if I do that I do get an emoji. `ð` is apparently `\xF0`.

All the more reason to have a comment that explains what it's meant to be :)

Reply:
The expectation has the actual emoji, and not an escape. Isn't that displayed correctly for you?

Reply:
Ah, true. It's there. I focused on the first one which has escapes in both places :)